ci/upload-docs.sh

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env bash
+#
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
+  echo "Skipping document generation as it is disabled for pull requests."
+  exit 0
+fi
+
+if [ "${GENERATE_DOCS:-}" != "yes" ]; then
+  echo "Skipping document generation as it is disabled for this build."
+  exit 0
+fi
+
+if [ "${TRAVIS_BRANCH:-}" != "master" ]; then
+  echo "Skipping document generation as it is disabled for non-master branches."
+  exit 0
+fi
+
+# The usual way to host documentation in ${GIT_NAME}.github.io/${PROJECT_NAME}
+# is to create a branch (gh-pages) and post the documentation in that branch.
+# We first do some general git configuration:
+
+# Clone the gh-pages branch into the doc/html subdirectory.
+REPO_URL=$(git config remote.origin.url)
```

Comment:
What happens if origin is `git@github.com:user/repo.git`?

In that case, `REPO_REF` is the same thing (since it just removes the `https://` prefix), but then how does this work at the end of the script?

```
git push https://${GH_TOKEN}@${REPO_REF} gh-pages
```

It's going to turn into something like:

```
git push https://${GH_TOKEN}@git@github.com:user/repo.git gh-pages
```

which I am guessing won't work well.

This script needs to handle both URL formats for `remote.origin.url`:

* `https://github.com/user/repo.git`
* `git@github.com:user/repo.git`

or please explain to me how it will handle the second case (if it does already), as I am not seeing it.

Reply:
Yes, the script does not work with `git@github:user/repo.git`.  The script runs is intended for use inside Travis.  Travis always checks out with `https://` (pretty sure, could be wrong about that, but pretty sure).  I would argue that the script does not *need* to handle the `git@github:...` case.   But I see how the error messages would be annoying, and how somebody might want to use the script from the command-line to test things out.

I do not think we want to spend a terrible amount of time making this case perfect, it is not the normal use for a script in the `ci/` subdirectory.  The user is already simulating the Travis environment: they did set TRAVIS_PULL_REQUEST, TRAVIS_BRANCH, DISTRO, DISTRO_VERSION, and just finished a build into the cached-${DISTRO}-${DISTRO_VERSION}:tip image.  Asking them to clone the code with `https://`, just like travis does, does not seem like a stretch.

Anyway, improved the code to handle this case better. Let me know what you think.

Reply:
SGTM, thanks for the explanation.